### PR TITLE
Improve Tab system

### PR DESCRIPTION
**Pull request by not-nef**

### CODE REVIEW

**Comment by not-nef** [comment]
@Moosems have fun implementing the close buttons

**Comment by Moosems** [comment]
> @Moosems have fun implementing the close buttons

I'll get a move on that

**Comment by Moosems** [comment]
I'm trying to work on it but it probably won't be able to be completed soon. I need to be a bit more familiar with how you set up the system so I'll let you know when I've got something working

**Comment by Moosems** [comment]
@not-nef I would go ahead with the pr while me and @sumeshir26 work on updating it. We can add the closing of tabs in that update which gives room for you to continue working

**Comment by not-nef** [comment]
Alright!

**Comment by Moosems** [comment]
If @akuli is ok with it we might also want to cheat and get some help from his tab system

**Comment by Akuli** [comment]
Feel free to borrow / get inspired / copypaste anything from my projects :) Most of them are licensed under the MIT license.
- If you copy/paste, you need to copy the first few lines of my LICENSE file, as the LICENSE file explains.
- If you don't copy/paste, you don't have to credit me in any way, but I would appreciate a comment along the lines of `# inspired by https://github.com/Akuli/porcupine/`

**Comment by Moosems** [comment]
Thank you so much Akuli. Your work really is incredible

**Comment by Akuli** [comment]
Looking at the conversation here a bit more, you are most likely interested in how the close buttons are implemented in Porcupine. I could just tell you the file name, but I think it's better to explain how I myself would go about finding it, so that you can similarly find other things by yourself:
- The close button obviously uses an image. Let's find all images by listing all files (`git ls-files`) and filtering the output to only contain PNG images (`| grep png`). Tkinter also supports gifs, but Porcupine doesn't use gif images.
    ```
    $ git ls-files | grep png
    docs/no-plugins.png
    porcupine/images/closebutton_dark.png
    porcupine/images/closebutton_light.png
    porcupine/images/pause_dark.png
    porcupine/images/pause_light.png
    porcupine/images/resume_dark.png
    porcupine/images/resume_light.png
    porcupine/images/stop.png
    porcupine/images/venv.png
    resource-hacker.png
    screenshot.png
    windows-defender.png
    ```
    Guessing from the file names, it is `closebutton_light.png` or `closebutton_dark.png`.
- Let's search for where the file names are used:
    ```
    $ git grep closebutton_dark.png
    ```
    No results. Maybe the `_dark` or `_light` suffix is added dynamically when Porcupine runs, and maybe that's why it's not mentioned as is in the code? Let's search for just `closebutton`:
    ```
    $ git grep closebutton
    ```
    This brings up several matches, and you can get a pretty good idea of how the image file ends up in the notebook just by looking at the output.

Instead of searching for an image, you would typically search for text that appears in the GUI.

**Comment by not-nef** [comment]
Man i love you ❤️

**Comment by not-nef** [comment]
@Moosems Hows it going?

**Comment by Moosems** [comment]
@not-nef Not great. I haven't had any time to code the past few days and am pretty busy until monday

**Comment by Moosems** [comment]
@not-nef Do we really need the buildtabs() function anymore?

**Comment by not-nef** [comment]
Not if you find another way of doing it

**Comment by Moosems** [comment]
I have some ideas :) I'll get to work on the tab system today

**Comment by Moosems** [comment]
Actually, I can just do what I did forget JDE

**Comment by not-nef** [comment]
@Moosems Any progress?

**Comment by Moosems** [comment]
Kinda. I need to redo a bit on my end and have been way busier than I should have been. In general what we need to do is just actually use multiple text boxes but to remove lag and remove to need for `buildtabs()` we just empty the text box when the tab is changed and add the text in the tab changed to

**Comment by not-nef** [comment]
Ill leave that to you. Take your time, i will wait.

**Comment by Moosems** [comment]
Honestly if you want work to get done its better not to wait on me and keep working so that I can just add my additions at some point

**Comment by not-nef** [comment]
I want to do one feature at a time

**Comment by Moosems** [comment]
@not-nef Can you help me work on this? Redoing this stuff gets really big really fast

**Comment by not-nef** [comment]
The problem is i have no idea how to implement the close buttons

**Comment by not-nef** [comment]
Ok. Can you tell me what i should do?

**Comment by Moosems** [comment]
First is just set it up to use individual text widgets and then remove buildtabs()

**Comment by Moosems** [comment]
Because of buildtabs(). What we can do is remove the text in it when the tab is switched and that removes the need to get rid of them all. It also won't hurt performance which makes this a great alternative

**Comment by Moosems** [comment]
buildtabs() looks janky when it builds the tabs

**Comment by not-nef** [comment]
> Because of buildtabs(). What we can do is remove the text in it when the tab is switched and that removes the need to get rid of them all. It also won't hurt performance which makes this a great alternative

Still dont see how that would require a textwidget for every tab

**Comment by Moosems** [comment]
Because if we use individual textwidgets it removes the need for buildtabs()

**Comment by not-nef** [comment]
I think we could eliminate buildtabs without requiring multiple texts

**Comment by Moosems** [comment]
How so?

**Comment by not-nef** [comment]
So lets say we have a function that gets called everytime a tab gets clicked. Lets call it `switchtab()`. So when switchtab gets called, it gets the index of the selected tab, and writes tabs[index][1] in the text.

**Comment by not-nef** [comment]
Tabs will get added in `new()`

**Comment by not-nef** [comment]
@Moosems i dont know how, but i managed to remove buildtabs

**Comment by Moosems** [comment]
Perfect! Great job!

**Comment by not-nef** [comment]
@Moosems Can you add the close buttons please? I still have no idea how to.

**Comment by sumeshir26** [comment]
Maybe use a image with compound right?

**Comment by Moosems** [comment]
My thoughts exactly

**Comment by not-nef** [comment]
But how would we detect if the user clicks on it? And how would we know which tab he wants to close?

**Comment by sumeshir26** [comment]
> But how would we detect if the user clicks on it? And how would we know which tab he wants to close?

Maybe using `bind`?

**Comment by Moosems** [comment]
That's what Akuli's Porcupine does

**Comment by not-nef** [comment]
What do you want to bind here?

**Comment by Moosems** [comment]
The image itself that's being compounded

**Comment by not-nef** [comment]
I wanna see you bind a parameter

**Comment by Moosems** [comment]
@not-nef Challenge accepted

**Comment by Moosems** [comment]
@not-nef Where is the tab add and tab destroy function? Does the open_file function not rely on these? I would recommend having the open_file function call on the tab create function and then just insert the text into that tab

**Comment by not-nef** [comment]
@Moosems Create tab: 

`v.tabbar.add(tkinter.Frame(), *args)`

**Comment by Moosems** [comment]
If I edit text in there to be `example` it stills says Untitled

**Comment by Moosems** [comment]
Ahh fixed it

**Comment by Moosems** [comment]
@not-nef Ok so I am trying to compound the image to the tab but every time I add a tab it removes the image after a second

**Comment by not-nef** [comment]
Push the code pls

**Comment by Moosems** [comment]
Pushed

**Comment by not-nef** [comment]
Will go to my pc give me a sec

**Comment by Moosems** [comment]
Try creating a new tab after it opens

**Comment by not-nef** [comment]
@Moosems the tab doesnt have an image at all

**Comment by not-nef** [comment]
and i have no idea why

**Comment by not-nef** [comment]
@Moosems i fixed your code

**Comment by Moosems** [comment]
Why is there an os import?

**Comment by not-nef** [comment]
I was checking if the path was correct by doing os.path.isfile

**Comment by not-nef** [comment]
I removed the import in the commit after the fix

**Comment by Moosems** [comment]
Can you create a function to remove a tab?

**Comment by not-nef** [comment]
@Moosems Hows it going?

**Comment by Moosems** [comment]
Not so great because the forget function brings up an error

**Comment by not-nef** [comment]
What error? Maybe i can help

**Comment by not-nef** [comment]
@Moosems Hows it going?

**Comment by not-nef** [comment]
I thought it `shouldn't be too difficult`

**Comment by Moosems** [comment]
I forgot to send the error message

```yaml
Exception in Tkinter callback
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/tkinter/__init__.py", line 1921, in __call__
    return self.func(*args)
  File "/Users/Moosems/Desktop/txt2/src/tabmanager.py", line 73, in opentab
    v.tabselected = v.tabbar.index(v.tabbar.select())
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/tkinter/ttk.py", line 866, in index
    return self.tk.getint(self.tk.call(self._w, "index", tab_id))
_tkinter.TclError: Invalid slave specification
```

**Comment by not-nef** [comment]
Can you push the code please?

**Comment by not-nef** [comment]
when does the error occur for you? i cant find it

**Comment by Moosems** [comment]
Line 119 (when you close the tab using the button)

**Comment by not-nef** [comment]
Strange, no error for me

**Comment by Moosems** [comment]
Very odd

**Comment by Moosems** [comment]
I know, I was going to make a check that it's on the image but with the error I can't even get to that point

**Comment by not-nef** [comment]
Mac issue maybe?

**Comment by Moosems** [comment]
Maybe. It seems more like a tkinter issue

**Comment by not-nef** [comment]
@Moosems Did you fix it?

**Comment by Moosems** [comment]
... No?

**Comment by Moosems** [comment]
@not-nef its an error with getting the index of the current tab

**Comment by not-nef** [comment]
The error looks like you specify a tab that doesnt exist

**Comment by not-nef** [comment]
I have an idea

**Comment by not-nef** [comment]
You are trying to get the index of the selected tab. But what if that tab doesnt exist?

**Comment by not-nef** [comment]
Check if the function gets called before the first tab gets created

**Comment by not-nef** [comment]
I am currently on vacation and dont have access to a pc

**Comment by Moosems** [comment]
> Check if the function gets called before the first tab gets created

I mean technically no because you don't call the tab creation function when starting

**Comment by not-nef** [comment]
Check if the function gets called at any point where no tab exists (if you close the last one for example)

**Comment by Moosems** [comment]
I think you misunderstand

**Comment by not-nef** [comment]
To me the error looks like the `tabbar.select()` function (which returns the selected tab) returns a slave that doesnt exist, so the `tabbar.index()` function throws an error. And the only explaination that sounds plausible to me is that the function (`tabbar.index(tabbar.select())`) gets called at a point where no tab exists. But of i think about it, `tabbar.select()` should also throw an error.

**Comment by not-nef** [comment]
Something that would support my suspicion is that the error comes from the index function. See the [ttk source code](https://github.com/python/cpython/blob/3.10/Lib/tkinter/ttk.py) at line 866

**Comment by not-nef** [comment]
That is what it looks like to me. So how do you see it?

**Comment by Moosems** [comment]
@not-nef it only happens when it's the first tab. Works with an if statement not allowing the closing of tab 0

**Comment by not-nef** [comment]
Dude this is exactly what im saying. The function (`v.tabselected = v.tabbar.index(v.tabbar.select())`) gets called everytime a tab gets closed. So if you close the only tab that exists, the tabbar.select() function returns nothing because there is no tab that can be selected, so the index function returns an error. Do you finally understand now?

**Comment by not-nef** [comment]
Fix:

```
try:
    v.tabselected = v.tabbar.index(v.tabbar.select())
except:
    v.root.destroy()
```

As i said, i cant fix it because i am on vacation, but you can do it if you want.

**Comment by not-nef** [comment]
@Moosems So what is your plan now?

**Comment by not-nef** [comment]
Btw i added a project board, can you see it?

**Comment by sumeshir26** [comment]
Nope, cant see it

**Comment by not-nef** [comment]
How about now?

**Comment by Moosems** [comment]
Honestly @not-nef the current implementation is so filled with confusion that I would personally refactor it such that it is more ash to read but since it's your project I don't want to make big changes without your permission

**Comment by not-nef** [comment]
Go ahead ;) i agree the code sucks

**Comment by not-nef** [comment]
@Moosems Do you have a plan on how to implement the closing of tabs? Because currently clicking on a tab will close it. We might have to ask akuli for help

**Comment by Moosems** [comment]
I was thinking we do exactly what akuli's does. If you click on the tab and the x and y coordinates touch the image then it closes, otherwise it stays open

**Comment by not-nef** [comment]
Sounds good, hows it goin?

**Comment by Moosems** [comment]
Honestly I would really appreciate it if you could make txt2 more readable. 90% of the time I use for coding when i'm trying to help is just spent trying to understand the layout and then 5% is error fixing

**Comment by not-nef** [comment]
Oh... ok. I think you are talking about tabmanager.py right?

**Comment by not-nef** [comment]
Tell me what you don't understand and i will try my best to add comments that explain it. Thats all i can currently do, because i am on vacation until Saturday.

**Comment by Moosems** [comment]
The whole var system. I would recommend not using global vars when possible and would recommend passing the notebook to functions or creating a better notebook class for the app

**Comment by not-nef** [comment]
Whats wrong with the vars system? If it is the `v.` prefix, just do `from vars import *`

**Comment by Moosems** [comment]
```python
from __future__ import annotations
from tkinter import Misc
from tkinter.ttk import Notebook
class foo(Notebook):
    def __init__(self, master: Misc, *args, **kwargs):
        self.master = master
        Notebook.__init__(self.master, *args, **kwargs)

    def foo(self):
        print("This is an example of a self contained class/function")
```

**Comment by not-nef** [comment]
Yeah ok you can transition to that if you want

**Comment by Moosems** [comment]
Trust me, it'll make life so much easier 😉

**Comment by Moosems** [comment]
I'll get to it

**Comment by not-nef** [comment]
Btw can you guys see the project board now?

**Comment by sumeshir26** [comment]
yes

**Comment by not-nef** [comment]
Great! 👍

**Comment by not-nef** [comment]
@Moosems Any progress?

**Comment by not-nef** [comment]
@Moosems No?

**Comment by not-nef** [comment]
@Moosems i decided to keep the vars system for the moment (i am going to rework it soon) and continued work on this. Its almost done! Just have to fix some bugs and do a big code cleanup.

**Comment by not-nef** [comment]
@Akuli i copied a few lines of code from porcupine for this but i didnt commit it yet. Just want to get green light from you. I will of course credit you

**Comment by not-nef** [comment]
no way i actually finished it

**Comment by HuyHung1408** [comment]
congrats on that :)

**Comment by Akuli** [comment]
Go ahead :)

File: src/tabmanager.py [hunk 1]
```diff
@@ -1,161 +1,137 @@
...
         except pygments.util.ClassNotFound: lexer = pygments.lexers.TextLexer
-        lexer = "pygments.lexers." + str(lexer).split(".")[-1].removesuffix(">")
+        lexer = "pygments.lexers." + str(lexer).split(".")[-1].removesuffix(">").removesuffix("'")
         v.textwidget.text._set_lexer(eval(lexer))
         try: v.textwidget.redraw()
-        except: pass
\ No newline at end of file
+        except: pass
+
+# The following two functions are property of Akuli
```
Comment (Akuli) [inline src/tabmanager.py:98]: Not really property of me, as they are now mostly written by you now:

```suggestion
# The following two functions contain code copied from https://github.com/Akuli/porcupine
```

**Comment by not-nef** [comment]
Thanks a lot akuli!

**Pull request merged**